rmw_fastrtps_shared_cpp/src/types/topic_cache.hpp

```diff
@@ -0,0 +1,236 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TYPES__TOPIC_CACHE_HPP_
+#define TYPES__TOPIC_CACHE_HPP_
+
+#include <algorithm>
+#include <iterator>
+#include <map>
+#include <set>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "fastrtps/participant/Participant.h"
+#include "fastrtps/rtps/common/Guid.h"
+#include "fastrtps/rtps/common/InstanceHandle.h"
+#include "rcutils/logging_macros.h"
+
+typedef eprosima::fastrtps::rtps::GUID_t GUID_t;
+
+class TopicCache;
+std::ostream & operator<<(
+  std::ostream & ostream,
+  const TopicCache & topic_cache);
+/**
+ * Topic cache data structure. Manages relationships between participants and topics.
+ */
+class TopicCache
+{
+private:
+  typedef std::map<GUID_t,
+      std::unordered_map<std::string, std::vector<std::string>>> ParticipantTopicMap;
+  typedef std::unordered_map<std::string, std::vector<std::string>> TopicToTypes;
+
+  /**
+   * Map of topic names to a vector of types that topic may use.
+   * Topics here are represented as one to many, DDS XTypes 1.2
+   * specifies application code 'generally' uses a 1-1 relationship.
+   * However, generic services such as logger and monitor, can discover
+   * multiple types on the same topic.
+   *
+   */
+  TopicToTypes topic_to_types_;
+
+  /**
+   * Map of participant GUIDS to a set of topic-type.
+   */
+  ParticipantTopicMap participant_to_topics_;
+
+  /**
+   * Helper function to initialize a topic vector.
+   *
+   * @param topic_name
+   */
+  void initializeTopic(const std::string & topic_name, TopicToTypes & topic_to_types)
```

Comment:
This is definitely a subjective style thing, but I find it weird to have a private method where the second argument is always one of the member variables.  It would make more sense to me to have the arguments to this be `topic_name` and `guid`, and have it initialize both (by accessing the member variables), but there may be some reason that this doesn't work.